### DISCUSSION TITLE
turbo_stream is a private method in after_login hook?

### DISCUSSION THREAD

**Discussion by xpressivecode** [janko/rodauth-rails#227] [Q&A]
In looking at the readme example, a simple implementation is throwing an error:

`private method 'turbo_stream' called for #<RodauthController:0x0000000003aef8>`

```
after_login do
    if rails_request.format.turbo_stream?
        rails_render(turbo_stream: rails_controller_instance.turbo_stream.refresh)
    end
end
```

Unfortunately the readme example doesn't show the use of the stream itself, so hopefully I'm just calling this wrong?
```
after_reset_password do
    if rails_request.format.turbo_stream?
        rails_render(turbo_stream: [...])
    end
end
```

I looked at some other discussions/issues and stumbled across someone calling it. which is what I'm basing my implementation off of
```
rails_render(
    turbo_stream: [
        rails_controller_instance.turbo_stream.replace('reset_password_button', partial: 'profile/reset_password'),
        rails_controller_instance.turbo_flashes
    ]
)
```

Did I forget to set something up? If the context matters. We need to show the login/create in a modal and after a successful login or account creation, we're trying to use a custom turbo action to refresh the existing page vs redirecting. It's all working great, until the point of sending our custom turbo stream action back.

rodauth-rails version 1.11.0
turbo-rails: 1.4.0

Thanks in advance!

**Comment by xpressivecode** [comment]
For the time being, I'm using.  I'm still curious as to what I was doing wrong in the original implementation given the readme.

```
if rails_request.format.turbo_stream?
    Turbo::StreamsChannel.broadcast_action_to(:dialog, action: :refresh)
end
```

**Comment by janko** [comment]
Yeah, I also understood from previous discussions that `turbo_stream` is a public method on Rails controllers. I guess it makes sense that it's private, because public methods are possible actions.

Regardless, I was considering exposing the `turbo_stream` method directly on the auth class, this is probably a good excuse then. I'll probably push that change shortly.

**Reply by janko** [reply]
@xpressivecode Just pushed the `turbo_stream` method to the main branch, and updated the example in the README. Would you mind testing it out?

Note that if you want to render a turbo stream response and halt further execution of Rodauth logic (which is your case), you'll want to use `return_response rails_render(...)`. Otherwise, if you're overriding a `*_view` method with a turbo stream response like in https://github.com/janko/rodauth-rails/discussions/214, then you don't need the `return_response` call.

**Reply by xpressivecode** [reply]
@janko I just tried the main branch out and everything works great.  It's a really clean implementation.

```
after_login do
    remember_login

    if rails_request.format.turbo_stream?
        set_notice_flash login_notice_flash

        return_response rails_render(turbo_stream: turbo_stream.refresh)
    end
end
```

Thanks, really appreciate the quick turn around on this!
